### PR TITLE
[GH-1719]: Move from webpack to Vite

### LINKED ISSUE #1719: Migrate UI from Webpack to Vite
### Change

As Vite is faster and does the job of webpack+other dev dependencies, we should move to it to improve developer experience, and final bundle size smaller.

### Are you interested in contributing a pull request for this task?

Yes

### PR DESCRIPTION
**Pull request by kinow**
GitHub issue resolved #1719

Pull request Description:

Moving from webpack to Vite for faster build, simpler configuration, faster dev server, and build.

Good refs:

- https://blog.boot.dev/javascript/migrating-vue-webpack-to-vitejs/
- https://vitejs.dev/

Time on `main`:

```
build: 35 seconds
test:unit 9 seconds 
test:e2e 57 seconds
coverage:unit 9 seconds
coverage:e2e 60 seconds
lint 4 seconds
serve (same as dev in vite branch) 18 seconds
mvn clean test in the UI project: 01:41 min
mvn -Pdev clean install  06:23 min (shortest of 2 runs)
```

Time after this change:

```
build 15-20 seconds
test:unit 4-5 seconds
test:e2e 40-45 seconds
coverage:unit 7 seconds
coverage:e2e 42-55 seconds
lint 2 seconds
dev 1 second
mvn clean test in the UI project: 01:16 min
mvn -Pdev clean install  05:52 min (shortest of 2 runs)
```

No extra tests to include. I did some manual testing from Eclipse with the code generated with Vite, and ran several times the commands. Cypress has been updated from 10 to 12. Maybe that brittle test has been fixed too :crossed_fingers: 

----

 - [ ] Tests are included.
 - [x] Documentation change and updates are provided for the [Apache Jena website](https://github.com/apache/jena-site/)
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx or JENA-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: jena-fuseki2/jena-fuseki-ui/package.json [hunk 1]
```diff
@@ -35,22 +35,15 @@
...
     "@cypress/code-coverage": "^3.10.0",
     "@cypress/vue": "^4.2.0",
     "@cypress/webpack-preprocessor": "^5.12.2",
     "@istanbuljs/nyc-config-babel": "^3.0.0",
-    "@vue/cli-plugin-babel": "~5.0.8",
-    "@vue/cli-plugin-e2e-cypress": "^5.0.8",
-    "@vue/cli-plugin-eslint": "~5.0.8",
-    "@vue/cli-plugin-unit-mocha": "~5.0.8",
-    "@vue/cli-service": "~5.0.8",
+    "@vitejs/plugin-vue": "^4.0.0",
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-ui/package.json:42]: Less dependencies as Vite does many of what these other tools do (and has been adopted by devs doing Vue, React, Angular, ThreeJS, so now it's stable and future-prone too :crossed_fingers: )

File: jena-fuseki2/jena-fuseki-ui/src/views/dataset/Upload.vue [hunk 2]
```diff
@@ -186,9 +186,9 @@
 </template>
 
 <script>
-import Menu from '@/components/dataset/Menu'
+import Menu from '@/components/dataset/Menu.vue'
 import FileUpload from 'vue-upload-component'
-import JenaTable from '@/components/dataset/JenaTable'
+import JenaTable from '@/components/dataset/JenaTable.vue'
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-ui/src/views/dataset/Upload.vue:191]: Vite does not try to automatic guess what's the extension of imports. So we need to specify it.

File: jena-fuseki2/jena-fuseki-ui/package.json [hunk 3]
```diff
@@ -62,12 +55,14 @@
...
     "istanbul-lib-coverage": "^3.2.0",
+    "jsdom": "^21.0.0",
     "json-server": "^0.17.0",
     "nodemon": "^2.0.20",
     "nyc": "^15.1.0",
     "sass": "^1.54.9",
-    "sass-loader": "^13.0.2",
     "sinon": "^14.0.0",
+    "vite": "^4.0.4",
+    "vitest": "^0.27.1",
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-ui/package.json:65]: Vitest is recommended in Vue3 docs https://vuejs.org/guide/scaling-up/testing.html#unit-testing

File: jena-fuseki2/jena-fuseki-ui/package.json [hunk 4]
```diff
@@ -62,12 +55,14 @@
     "express": "^4.18.1",
     "istanbul-instrumenter-loader": "^3.0.1",
     "istanbul-lib-coverage": "^3.2.0",
+    "jsdom": "^21.0.0",
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-ui/package.json:58]: For vitest/tests.

File: jena-fuseki2/jena-fuseki-ui/package.json [hunk 5]
```diff
@@ -35,22 +35,15 @@
...
-    "@vue/cli-plugin-e2e-cypress": "^5.0.8",
-    "@vue/cli-plugin-eslint": "~5.0.8",
-    "@vue/cli-plugin-unit-mocha": "~5.0.8",
-    "@vue/cli-service": "~5.0.8",
+    "@vitejs/plugin-vue": "^4.0.0",
     "@vue/compiler-sfc": "^3.2.39",
     "@vue/eslint-config-standard": "^8.0.1",
     "@vue/test-utils": "^2.0.2",
     "babel-plugin-istanbul": "^6.1.1",
-    "chai": "^4.3.6",
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-ui/package.json:53]: Vitest has assertions. No need for chai anymore.

File: jena-fuseki2/jena-fuseki-ui/package.json [hunk 6]
```diff
@@ -35,22 +35,15 @@
     "vue-upload-component": "^3.1.2"
   },
   "devDependencies": {
-    "@babel/eslint-parser": "^7.19.1",
-    "@babel/register": "^7.18.9",
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-ui/package.json:39]: I think it should be fine to drop support to older browsers (IE, older Ffox, Chrome... Ffox and Chrome have supported newer JS for years now, nobody should be using IE... this reduces our build time & size - makes UI/JS development a lot easier too).

**Comment by kinow** [comment]
#TIL Vitest includes coverage, and it appears to be a lot faster than vue-test-utils+chai+nyc:

image

The coverage metrics also appear to be more useful as it includes more levels of coverage (not sure if we needed to configure that in nyc).

**Comment by afs** [comment]
Do you want to create an issue for this? (which can be auto-closed by the PR).

An issue is a little more findable and can pull together several work items if necessary.

**Comment by kinow** [comment]
> Do you want to create an issue for this? (which can be auto-closed by the PR).
> 
> An issue is a little more findable and can pull together several work items if necessary.

Sure thing. Consider it done. What about `jena-site`? Should we have an issue for each PR there too?

**Comment by afs** [comment]
We haven't been using issues for the site. They aren't enabled. This lighter weight process has been OK for contributions as far as I can see.

The code issues for the release description.

**Comment by kinow** [comment]
Work is all done, and apparently Fuseki UI is running with no major problems. I'm having a problem to dynamically control the coverage in e2e tests, but no other issues. Just pending to do some tests with Fuseki servlets, and get the numbers from `main` with webpack to compare :+1: should be ready for review in the next day(s).

**Comment by kinow** [comment]
Done. Build should be a bit faster now :+1:

**Comment by afs** [comment]
It's faster!

I get the old error - but with more information about where:

First:
```
[INFO] [TESTS]   0 passing (31s)
[INFO] [TESTS]   4 failing
[INFO] [TESTS] 
[INFO] [TESTS]   1) datasets
[INFO] [TESTS]        Visits datasets page, also the application landing-page:
[INFO] [TESTS]      CypressError: `cy.visit()` failed trying to load:
[INFO] [TESTS] 
[INFO] [TESTS] http://localhost:8080/
[INFO] [TESTS] 
[INFO] [TESTS] We attempted to make an http request to this URL but the request failed without a response.
[INFO] [TESTS] 
[INFO] [TESTS] We received this error at the network level:
[INFO] [TESTS] 
[INFO] [TESTS]   > Error: ESOCKETTIMEDOUT
[INFO] [TESTS] 
[INFO] [TESTS] Common situations why this would fail:
[INFO] [TESTS]   - you don't have internet access
[INFO] [TESTS]   - you forgot to run / boot your web server
[INFO] [TESTS]   - your web server isn't accessible
[INFO] [TESTS]   - you have weird network configuration settings on your computer
...
```

```
[INFO] [TESTS]        Spec                                              Tests  Passing  Failing  Pending  Skipped  
[INFO] [TESTS]   ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
[INFO] [TESTS]   │ ✖  datasets.cy.js                           00:30        9        -        3        -        6 │
[INFO] [TESTS]   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
[INFO] [TESTS]   │ ✖  query.cy.js                              00:15        2        -        1        -        1 │
[INFO] [TESTS]   └────────────────────────────────────────────────────────────────────────────────────────────────┘
[INFO] [TESTS]     ✖  2 of 2 failed (100%)                     00:45       11        -        4        -        7  
```

**Review by afs — APPROVED** [review]

**Comment by kinow** [comment]
> It's faster!

Hooray! And we will have less direct dependencies to worry (not sure about number of transitive deps).

> I get the old error after 2 clean runs - with more information about where:
> 

Hmmmm... I wonder if the tests are running before the fake Fuseki is up and serving. I think I can simulate a very slow startup to confirm if the Cypress process is waiting on the fake Fuseki (`json-server`) to start up. A good result of updating the version of Cypress, I guess.

Thanks @afs!

Merging.

**Pull request merged**